# An interactive overview of QIIME 2

In the previous chapter I introduced you to some concepts that will help you learn QIIME 2 more quickly. Now that we're through that, let's jump in and do something really quickly with QIIME 2. In this chapter I'll present a workflow, which you should follow along with on your computer if possible, that will demultiplex sequences, apply quality control and build a feature table, and build five different interactive visualizations. The interactive that we'll build in this chapter are the first ones that I generate when working with a new data set. They will help you to judge the quality of your sequencing run, and give you an initial view of the alpha diversity, beta diversity, and taxonomic composition of your samples. Before you can apply these steps to your own data, you'll also need to learn how to import your data into a QIIME 2 artifact. This will be covered in the next chapter. 

## The data

For the purpose of getting a very quick overview of QIIME 2, we're going to work with a desert soil microbiome data set from the Atacama Desert in northern Chile. The Atacama desert is one of the most arid locations on the Earth. In many years it receives no rain at all. The data we'll use here is sampled along a gradient from the hyper-arid core of the desert to the western foothills of the Andes mountain range, where there is more moisture resulting from snowmelt in the mountains. 

I have a few hypotheses going into this analysis that we can validate when we analyze the data. First, I expect to see dessication resistant organisms dominating the samples in the hyper-arid core of the desert. Second, I hypothesize that there will be fewer types of microbes, whether I look at the genus level or higher taxonomic levels, in the hyper-arid samples than in the less arid samples. Third, I expect the samples from the hyper-arid sites to be more similar to each other in overall composition than they are to samples from the less arid sites. I'm formulating these ideas now, before I start analyzing the data, so I have an idea of what I expect to see. After analyzing the samples, I'll have a good idea of whether the data meets my expectations or whether something surprizing is going on.

We'll begin this analysis by downloading a QIIME 2 artifact that contains our demultiplexed sequence data. (As mentioned above, in the next chapter we'll cover generating this artifact from your sequencing data by importing it.) Do this as follows:

```bash

wget \
  -O "sample-metadata.tsv" \
  "https://data.qiime2.org/2020.11/tutorials/atacama-soils/sample_metadata.tsv"

wget \
  -O "emp-paired-end-sequences.qza" \
  "https://docs.qiime2.org/2020.11/data/tutorials/atacama-soils/emp-paired-end-sequences.qza"

```  

You'll see some output on the screen, and you should ultimately see that these two files were downloaded successfully. 

```{warning}
If it seems like the download failed, remove any files that were generated by these commands and try to download them again. If you're still running into errors, you should feel free to post a question to the QIIME 2 Forum. In general, if you're confused about something in the book, or something isn't working for you, the QIIME 2 Forum is the best place to get help. Start by searching to see if anyone has asked a similar question that has been answered. If not, post your own question. It can be intimidating to post questions to online forums, but it can also be extremely helpful. Don't worry - the QIIME 2 community is very friendly and many of the people posting questions there have little or no experience using command line software before trying QIIME 2, so even if your question seems basic, you can rest assured that someone has had or will have almost the same question you're asking. 
```

The commands that you ran downloaded two files that will get us started on this analysis. The first, sample-metadata.tsv, is a tab-separated text file (the `.tsv` file extension tells you this). In QIIME 2, the sample metadata file maps sample identifiers to information about those samples. You should be able to open and view that file in a program such as Google Sheets (any text editor, such as NotePad, NotePad++, TextEdit, or TextMate, will also work, but the layout that will provided by a spreadsheet viewer will help you interpret the file contents). In this data set you'll see that sample metadata includes information such as the latitude and longitude where a sample was collected, whether that site was classified as hyper-arid, arid, or semi-arid. As you can imagine, that information is essential for letting us explore the hypotheses that I listed above. Sample metadata is absolutely essential to your analysis. We'll revisit this in a lot more detail a little bit later.

The second file that we downloaded above was our multiplexed sequence data. This is sequencing data as it came off an Illumina DNA sequencing instrument. At this stage, the sequences are all grouped together - they are not grouped by sample. Instead, during sample preparation (PCR specifically), a short DNA sequence referred to as an index or a barcode was added to sequences on a per-sample basis. For example, sample-1 might be assigned the barcode TGACCGTACGTA, and sample-2 might be assigned the barcode TGGTAGACCCGT. All sequences derived from sample-1 will have the sample-1 barcode associated with them, and all sequences derived from sample-2 will have the sample-2 barcode associated with them. One piece of sample metadata in our sample metadata file is the barcode sequence associated with each sample, and we can use the `q2-demux` plugin to assign sequences to samples. 

As a first analysis step with QIIME 2, demultiplex the sequence data with the following command:

```
qiime demux emp-paired \
  --i-seqs emp-paired-end-sequences.qza \
  --m-barcodes-file sample-metadata.tsv \
  --m-barcodes-column barcode-sequence \
  --p-rev-comp-mapping-barcodes \
  --o-per-sample-sequences demux.qza \
  --o-error-correction-details demux-log.qza
```

Since this is one of the first QIIME 2 commands that you've run (at least while reading this book) let's take a look at it in some detail. As we look at this command, I'm going to assume that you either have some command line software experience, or that you've read _the introductory chapter in this book on using the command line_. There are four components to notice on the first line of this command, and as is typically the case when running command line software, the command's components are separated by spaces. The first space-separated component is `qiime`. This tells your computer that the QIIME 2 program should be run. The second space-separated component is `demux`. This is the name of the q2-demux plugin, as far as QIIME 2 is concerned, so this tells QIIME 2 that you want to use the `q2-demux` plugin. The third space-separated component is `emp-paired`, which is an action in the q2-demux plugin. This action demultiplexes paired-end sequencing data that is formatted as is typical for Earth Microbiome Project (EMP) sequenced data (more on that later). As mentioned in the previous chapter, if you'd like to learn about this action, you could run the command `qiime demux emp-paired --help`, which will print help text to the screen. Finally, the `\` at the end of the line is worth mentioning now. When you're working on the command line, line breaks (i.e., the character that is received when you press the _Return_ key on your keyboard) signifies that you have finished entering the command and that the command should now be executed (i.e., run). Since QIIME 2 commands can be long, it's helpful when documenting them to split them over multiple lines so that they can be read without the reader having to scroll to the right. If the documentation is being presented in a non-interactive medium, such as a printed book, the formatting of the command could actually be misleading - for example, it could look like the command should be split across multiple lines. Splitting long commands across multiple lines also therefore helps to ensure accuracy of the documentation. The `\` character here simply means that the terminal shouldn't interpret the line break that follows it as the end of the command, but rather that the command will continue on the next line. The same command, without the line breaks, would look like the following:

``` 
qiime demux emp-paired --i-seqs emp-paired-end-sequences.qza --m-barcodes-file sample-metadata.tsv --m-barcodes-column barcode-sequence --p-rev-comp-mapping-barcodes --o-per-sample-sequences demux.qza --o-error-correction-details demux-log.qza
```

That command will do the exact same thing as the one above. I think you'll agree though that the first variant, where the command is split across multiple lines, is easier to read. 

After we specify the program (`qiime`), the plugin (`demux`), and the action (`emp-paired`) that we want to run, we provide options to that action to tell it what files we want it to operate on, what we want it to do exactly, and what we want the outputs to be called. These components of the command are specified on the remaining lines. Let's look at those now. 

This action takes one QIIME 2 artifact as input, the mutliplexed sequences that we downloaded above. In the QIIME 2 command line interface, input artifacts are always specified with options beginning with `--i-`. This action also takes a metadata file and a metadata column, both of which are specified with options beginning with `--m-` (where the _m_ implies _metadata_). Here, these specify the file containing the metadata (that's the second file that we downloaded above) and column header in that file that contains the barcode sequences. If you have opened the `sample-metadata.tsv` file, you'll see that there is a column called `barcode-sequence`. This column could be called nearly anything (there are some rules for what column names are allowed that we'll cover later). As long as you specify the column name correctly when you call this action, it will know where to find the barcodes. Input and metadata parameters, in general, are files that you're providing to a QIIME 2 action for actions to be performed on. They will never be modified by QIIME 2 - they are only viewed or read. In this command we're also providing one parameter, specified by the option beginning with `--p-`. Parameters modify the behavior of an action, and parameters that don't take any values (such as this one) are often referred to as _flags_. We'll see examples of parameters that take a value later in this chapter. Finally, this command will generate two files as output, and we specify the paths (absolute or relative) where we want to store those files with output options. Output options always begin with `--o-` in QIIME 2. The first output created by this command was `demux.qza`, which contains our demultiplexed sequences (or sequences that have been grouped by sample). The second output created by this command is `demux-log.qza`. This is a log file that contains some details about this action. 

```
qiime demux summarize \
  --i-data demux-subsample.qza \
  --o-visualization demux-subsample.qzv
```

```
qiime dada2 denoise-paired \
  --i-demultiplexed-seqs demux.qza \
  --p-trim-left-f 13 \
  --p-trim-left-r 13 \
  --p-trunc-len-f 150 \
  --p-trunc-len-r 150 \
  --o-table table.qza \
  --o-representative-sequences rep-seqs.qza \
  --o-denoising-stats denoising-stats.qza
```

```
qiime feature-table summarize \
  --i-table table.qza \
  --o-visualization table.qzv \
  --m-sample-metadata-file sample-metadata.tsv

qiime feature-table tabulate-seqs \
  --i-data rep-seqs.qza \
  --o-visualization rep-seqs.qzv

qiime metadata tabulate \
  --m-input-file denoising-stats.qza \
  --o-visualization denoising-stats.qzv
```


