# An interactive overview of QIIME 2

In the previous chapter I introduced you to some concepts that will help you learn QIIME 2 more quickly. Now that we're through that, let's jump in and do something quick with QIIME 2. In this chapter I'll present a workflow, which you should follow along with on your computer if possible. In this workflow we'll download some real sequencing data and associated sample metadata, and apply some QIIME 2 methods and visualizers to it to see if it supports a hypothesis about the biological system being studied. The interactive visualizations that we'll build in this chapter are some of the first ones that I generate when working with a new data set. They will help you to judge the quality of your sequencing run, and give you an initial view of the alpha diversity of your samples. Before you can apply these steps to your own data, you'll also need to learn how to import your data into a QIIME 2 artifact. This will be covered in {ref}`importing`. 

**TODO: my cross references aren't working - see "importing" directly above this line.**

## The study

For the purpose of getting a very quick overview of QIIME 2, we're going to work with a desert soil microbiome data set that I was involved with generating to study the Atacama Desert in northern Chile {cite}`Neilson2017-bz`. The Atacama Desert is one of the most arid locations on the Earth: in many years the hyper-arid core of the desert receives no rain at all. The data we'll use here is sampled along a gradient from the hyper-arid core of the desert to the western foothills of the Andes mountain range, where there is more moisture resulting from snowmelt in the mountains. 

I have a few hypotheses going into this analysis that we can test when we analyze the data. I'm formulating these hypotheses now, before I start analyzing the data, so I have an idea of what I expect to see. After analyzing the samples, I'll have a good idea of whether the data supports my hypotheses or whether I'm surprised by the results. For the purpose of this introductory analysis I'm going to focus on one hypothesis. 

 * Hypothesis: I hypothesize that there will be fewer types of microbes in the hyper-arid samples than in the less arid samples. 

"Type of microbe" is a term that needs to be defined. What I mean by that is an arbitrary taxonomic grouping, such as genus, species, or strain. The data we'll work with in this chapter is 16S rRNA amplicon data. To keep the analysis in this chapter simple, we're not going to assign taxonomy to our amplicon sequences. The unit of taxonomy that we'll work with here is the highest resolution possible for a study based on amplicon data: the _amplicon sequence variant_, or ASV. Each unique 16S rRNA amplicon that we observe in this study will be treated as a different ASV, and thus a different "type of microbe". Due to the limits of resolution in a 16S rRNA amplicon study, this approximates genus or species but does not consistently represent a single named taxonomic level.

## Downloading the data

We'll begin this interactive overview by downloading a few files to work on. Do this as follows:

```bash

curl -sL \
  "https://data.qiime2.org/2020.11/tutorials/atacama-soils/sample_metadata.tsv" > \
  "sample-metadata.tsv"

curl -sL \
  "https://docs.qiime2.org/2020.11/data/tutorials/atacama-soils/demux.qza" > \
  "demux.qza"
```

You'll see some output on the screen, and you should ultimately see that these two files were downloaded successfully. 

```{warning}
If it seems like the download failed, remove any files that were generated by these commands and try to download them again. If you're still running into errors, you should feel free to post a question to the [QIIME 2 Forum](https://forum.qiime2.org). In general, if you're confused about something in the book, or something isn't working for you, the QIIME 2 Forum is the best place to get help. Start by searching to see if anyone has asked a similar question that has been answered. If not, post your own question. It can be intimidating to post questions to online forums, but it can also be extremely helpful. Don't worry - the QIIME 2 community is very friendly and many of the people posting questions there had little or no experience using command line software before trying QIIME 2. Even if your question feels hopelessly basic, you can rest assured that someone has had or will have almost the same question you're asking. 
```

The two `curl` commands that you just ran each downloaded a file that we'll use to get started on this analysis. The first, `sample-metadata.tsv`, is a tab-separated text file (the `.tsv` file extension tells you this). In QIIME 2, the sample metadata file maps sample identifiers to information about those samples. You should be able to open and view that file in a program such as Google Sheets (any text editor, such as NotePad, NotePad++, TextEdit, or TextMate, will also work, but the layout that will provided by a spreadsheet viewer will help you interpret the file contents). In this data set you'll see that sample metadata includes information such as the latitude and longitude where a sample was collected, whether that site was classified as hyper-arid, arid, or semi-arid. As you can imagine, that information is essential for letting us explore the hypotheses that I listed above. Sample metadata is absolutely essential to your analysis, a topic we'll revisit many times throughout this book.

The second file that we downloaded above is a QIIME 2 artifact containing our demultiplexed sequence data. This is sequencing data that has undergone some processing since it came off an Illumina DNA sequencing instrument. Specifically, the sequences have already been mapped to the samples they were observed in: a process referred to as demultiplexing. You can start a QIIME 2 analysis with data that is demultiplexed, as illustrated here, or data that is multiplexed. We'll discuss this more in {ref}`demultiplexing`. In this chapter we're starting with already demultiplexed data to keep this overview brief. 

## Viewing summaries of sequencing run quality and demultiplexing

As mentioned in the previous chapter, QIIME 2 artifacts such as the `demux.qza` file that you just downloaded are intermediary files in a QIIME 2 analysis. They're not meant to be directly viewed by humans. Generally when there is something important to view from a QIIME 2 artifact, such as a summary of the data, there will be a QIIME 2 visualizer that you can apply to that artifact. Some QIIME 2 visualizers are specific to a QIIME 2 artifact type, while others are more general purpose. We'll look at an example of each in this chapter.

The first visualizer that we'll use is one that is specific to visualizing summaries of demultiplexed sequence data with quality scores. In addition to a summary of the demultiplexing run, it provides information on the quality of sequence data in the form of interactive visualizations. You can run this command as follows:

```
qiime demux summarize \
  --i-data demux.qza \
  --o-visualization demux.qzv
```

This command should run successfully for you. You'll know it did if you see text printed to the terminal saying something like:

```
Saved Visualization to: demux.qzv
```

````{margin}
```{note}
[This video](https://t.co/eJbm03cnSa) on the QIIME 2 YouTube channel illustrates how to use QIIME 2 View. 
```
````

The next step is viewing this visualization. You can do this in a few ways, but the one I recommend right now is using [QIIME 2 View](https://view.qiime2.org). Open that page in your web browser now. You should see a big box on the screen where you can drop a QIIME 2 artifact or visualization to view it. Load the `demux.qzv` visualization by dragging and dropping it onto that box. A QIIME 2 visualization should load.


Notice that this visualization has two tabs within the page respectively labeled _Overview_ and _Interactive Quality Plot_. The _Overview_ tab provides a summary of the sequence counts per sample (i.e., the number of sequences that were obtained for each sample in the sequencing run) ({numref}`demux-1`). Because this is a paired-end sequencing run this is presented for both the forward and reverse reads. 

The median sequence per sample count is 7118. This is a bit lower than I typically like to see for a 16S Illumina run. I notice when looking at the details on the bottom of the page though that a lot of samples had very low sequence counts. Scroll down in the visualization on QIIME 2 View to see the number of sequences obtained for each sample. The very low sequences counts (less than a few thousand, though this is subjective) likely indicates an issue with DNA extraction or amplification of those samples. 

```{figure} ./images/demux-1.png
---
name: demux-1
---
Screenshot of `demux.qzv`.
```

The second tab in this visualization, _Interactive Quality Plot_, presents a summary of sequence quality scores (y-axis) by sequence position (x-axis). Open that tab in QIIME 2 View. A box plot is presented at each sequence position, where each box presents five values from [a seven-number summary](https://en.wikipedia.org/wiki/Seven-number_summary). As you hover over the boxes in that plot, the full seven-number summary will populate the table below the plot. The forward read quality plot, and a seven-number summary of the read quality at a single position, are presented in {numref}`demux-2`. This looks like a high quality sequencing run to me. 

```{figure} ./images/demux-2.png
---
name: demux-2
---
Plot illustrating sequence quality by sequence position for forward sequencing reads. This plot and other information are presented in the _Interactive Quality Plot_ tab of `demux.qzv`.
```

Spend a few minutes exploring this visualization, but don't worry too much about the details right now. We're going to discuss these plots, as well as demultiplexing in general, in more detail in other chapters. Remember, the goal of this chapter is just to do a few things quickly with QIIME 2.   

## Denoising the demultiplexed sequences

Since our sequence quality generally looks good, let's move on to the next step of our analysis: sequence quality control. This will determine which sequences we'll include in our downstream analyses. QIIME 2 has a few options for performing quality control. Here we'll use the DADA2 denoising approach {cite}`Callahan2016-ga`, which we'll access through the `q2-dada2` plugin. Go ahead and run this command, which will take up to about 10 minutes to complete.

```
qiime dada2 denoise-single \
  --i-demultiplexed-seqs demux.qza \
  --p-trim-left 13 \
  --p-trunc-len 150 \
  --o-table asv-table.qza \
  --o-representative-sequences asv-sequences.qza \
  --o-denoising-stats denoising-stats.qza
```

**TODO: Move detailed discussion of inputs, parameters, etc here since demux (where it was originally) is moving to the next chapter.**

As you can tell from the first line of this command, we're calling the program `qiime`, the plugin `dada2`, and the action `denoise-single`. This runs DADA2's single-end read denoiser, which may seem like a surprising choice given that I previously mentioned that this is paired end data. The reason that I choose to use the single end denoiser here is that these are short sequence reads for the region of the 16S rRNA that we're sequencing. The region is about 350 bases long, but the sequencing run was a 2x150 run (where the 2 implies that it was paired end, for two reads, and 150 implies that 150 bases were read in each direction). These reads are too short for paired end joining for this region of the 16S, so I'm not going to join them here. (If we were to apply paired end joining with the `denoise-paired` action, the reads that failed to join would be excluded from the results. This would result in a bias for organisms with shorter 16S sequences, which would likely have a confusing impact on our findings.)

Your `denoise-single` command may have finished running by now. If not, take a break for a few minutes. Stretch your legs, go for a walk around the block, or grab a glass of water. Waiting for a long-running job (e.g., the command that is currently running) to finish is a great opportunity to break up your screen time.

```{figure} https://imgs.xkcd.com/comics/compiling.png
---
height: 250px
name: xkcd-compiling
---
For programmers, waiting for code to compile is an excuse for a break. For data analysts, it's waiting for a long-running job to finish.
```

## Viewing a summary of the denoising run

The next visualizer that we'll use is a very powerful and general purpose visualizer that you should always keep in mind. It's the `tabulate` visualizer in the `q2-metadata` plugin. This might not seem very intuitive at the moment, but many QIIME 2 artifacts that provide some information on a per-sample basis can be used as if they are sample metadata. The power of this will become more clear in later chapters. We'll apply this to the log artifact that was generated as follows:

```
qiime metadata tabulate \
  --m-input-file denoising-stats.qza \
  --o-visualization denoising-stats.qzv
```

Load the resulting `.qzv` file with [QIIME 2 View](https://view.qiime2.org). You should see something that looks like {numref}`denoising-stats-1`. 

```{figure} ./images/denoising-stats-1.png
---
name: denoising-stats-1
---
Screenshot of `denoising-stats.qzv`.
```

## Viewing the ASV sequences and ASV table summary

Recall that the unit of taxonomy that we're working with in this chapter is the amplicon sequence variant (ASV). When DADA2 denoises the sequencing run, it defines the ASVs that were observed from the quality-controlled sequences. One of the outputs that it generates is the collection of the ASV sequences that were defined. In our example, that file is called  `asv-sequences.qza`. The `q2-feature-table` defines a special purpose visualizer for generating a human-readable list of the ASVs. You can run this as follows:

```
qiime feature-table tabulate-seqs \
  --i-data asv-sequences.qza \
  --o-visualization asv-sequences.qzv
```

Load `asv-sequences.qzv` using QIIME 2 View.  

```
qiime feature-table summarize \
  --i-table asv-table.qza \
  --o-visualization asv-table.qzv \
  --m-sample-metadata-file sample-metadata.tsv
```

These are referred to  one of the outputs it generates is a "feature table". This table has samples on one axis, ASVs on the other axis, and the values in the table are the number of times an ASV was observed in a sample. 

## Computing community richness at different depths of sequencing

```
qiime diversity alpha-rarefaction --i-table feature-table.qza --p-max-depth 10000 --m-metadata-file sample-metadata.tsv --o-visualization alpha-rarefaction.qzv
```

## Exercises

1. Use `qiime metadata tabulate` to generate a summary of the `demux-log.qza` file that was generated above. View the resulting `.qzv` with QIIME 2 View. (**TODO: this might be a little much for this stage of the analysis. It's also a big file, so it might overwhelm memory on the reader's system.**)

## List of works cited

```{bibliography} ../references.bib
:filter: docname in docnames
```