
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>1. Getting started with using QIIME 2 &#8212; q2book</title>
    
  <link rel="stylesheet" href="../_static/css/index.73d71520a4ca3b99cfee5594769eaaae.css">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.3da636dd464baa7582d2.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="2. An interactive overview of QIIME 2" href="overview-tutorial.html" />
    <link rel="prev" title="Preface" href="../front-matter/preface.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../index.html">
  
  <img src="../_static/logo.png" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">q2book</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../front-matter/preface.html">
   Preface
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Microbiome Bioinformatics with QIIME 2
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   1. Getting started with using QIIME 2
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="overview-tutorial.html">
   2. An interactive overview of QIIME 2
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="importing.html">
   3. Importing data into QIIME 2
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="demultiplexing.html">
   4. Demultiplexing a sequencing run
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="denoising.html">
   5. Denoising demultiplexed sequencing data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="taxonomy.html">
   6. Taxonomic annotation and analysis of sequences
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Understanding the algorithms
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../algorithms/pairwise-alignment.html">
   1. Pairwise sequence alignment
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../algorithms/database-searching.html">
   2. Sequence homology searching
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Developing with QIIME 2
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../developing/getting-started.html">
   1. Getting started with developing with QIIME 2
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../developing/first-plugin.html">
   2. Building a first plugin
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Back Matter
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../back-matter/biological-information.html">
   1. Appendix 1: Biological Information
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../back-matter/microbiome-methods.html">
   2. Appendix 2: A brief introduction to methods in microbiome science
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../back-matter/q2-origin-story.html">
   3. Appendix 3: History of the QIIME platform
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../back-matter/glossary.html">
   4. Glossary
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/using/getting-started.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#deploying-qiime-2">
   1.1. Deploying QIIME 2
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#using-qiime-2-on-your-personal-computer">
     1.1.1. Using QIIME 2 on your personal computer
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#using-qiime-2-on-a-cluster-computer">
     1.1.2. Using QIIME 2 on a cluster computer
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#using-qiime-2-on-the-amazon-cloud">
     1.1.3. Using QIIME 2 on the Amazon cloud
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#installing-qiime-2">
     1.1.4. Installing QIIME 2
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#qiime-2-archives">
   1.2. QIIME 2 archives
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#plugins-and-actions">
   1.3. Plugins and actions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#semantic-types">
   1.4. Semantic types
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#qiime-2-view">
   1.5. QIIME 2 View
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#getting-started">
   1.6. Getting started
  </a>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="getting-started-with-using-qiime-2">
<h1><span class="section-number">1. </span>Getting started with using QIIME 2<a class="headerlink" href="#getting-started-with-using-qiime-2" title="Permalink to this headline">¶</a></h1>
<p>This chapter will briefly introduce a few concepts that you should understand as you start using QIIME 2. This chapter is purposefully brief so you can get to the fun stuff soon. After reading this chapter you’ll understand some key QIIME 2 terminology that will be used throughout this book.</p>
<div class="section" id="deploying-qiime-2">
<h2><span class="section-number">1.1. </span>Deploying QIIME 2<a class="headerlink" href="#deploying-qiime-2" title="Permalink to this headline">¶</a></h2>
<p>The first thing you may be wondering as you prepare to learn QIIME 2 is where and how you’ll deploy it. QIIME 2 can be deployed on your personal computer (e.g., your laptop or desktop computer), a cluster computer such as one owned and maintained by your university, or on cloud computing resources such as the Amazon Web Services (AWS) Elastic Compute Cloud (EC2). In the following sections I describe these options for deploying QIIME 2, and I conclude with linking you to specific instructions on how to install QIIME 2. I recommend having a working deployment of QIIME 2 as you read this book so you can run the examples yourself.</p>
<div class="section" id="using-qiime-2-on-your-personal-computer">
<h3><span class="section-number">1.1.1. </span>Using QIIME 2 on your personal computer<a class="headerlink" href="#using-qiime-2-on-your-personal-computer" title="Permalink to this headline">¶</a></h3>
<p>Using QIIME 2 on your personal computer is a very convenient option, but may be too slow for some steps of your analysis workflow. Steps such as sequence quality control and taxonomic assignment may require more CPU or memory resources than are available on your personal computer. This might make some of those steps impossible to run, or might render your personal computer useless for other tasks while you’re waiting for a run to complete. That could take days or even weeks, depending on how powerful your personal computer is and how big the data set is that you’re working with.</p>
<p>A related option is that you could have a dedicated computer for running QIIME 2 analyses. This could be a server that your team owns and maintains, and that everyone on your team has access to. Then, if you need to let a job run for a long time, that machine can run for days or weeks without disrupting other work that you need to do on your personal computer. If you go with this option, it’s best to think about having that computer connected to a backup power supply in case of a power outage days into an analysis.</p>
<p>QIIME 2 can be installed natively on macOS or Linux personal computers, and on Windows personal computers that support the Windows Subsystem for Linux. QIIME 2 can also be used on personal computers through virtual machines. Virtual machines are a useful option if QIIME 2 can’t be installed natively on your computer, but should probably not be used otherwise as the overhead of running the virtual machine will reduce the CPU and memory resources that are available to QIIME 2.</p>
</div>
<div class="section" id="using-qiime-2-on-a-cluster-computer">
<h3><span class="section-number">1.1.2. </span>Using QIIME 2 on a cluster computer<a class="headerlink" href="#using-qiime-2-on-a-cluster-computer" title="Permalink to this headline">¶</a></h3>
<p>If you have access to a cluster computer, this is a great option for running QIIME 2. A cluster will typically have the resources needed to run QIIME 2 on large data sets. Typically you can reach out to your institution’s high performance computing or research computing office, and let them know that you need to use QIIME 2. They will have a process for having it installed on the system and made available for you to use.</p>
<p>One downside of using QIIME 2 on a cluster is that often you’ll only have command line (i.e., terminal) access to the cluster - not a graphical interface. This means that to view QIIME 2 results you’ll need to move them off that computer to your local machine for viewing. This isn’t a problem - just a minor inconvenience that you’ll need to get used to.</p>
</div>
<div class="section" id="using-qiime-2-on-the-amazon-cloud">
<h3><span class="section-number">1.1.3. </span>Using QIIME 2 on the Amazon cloud<a class="headerlink" href="#using-qiime-2-on-the-amazon-cloud" title="Permalink to this headline">¶</a></h3>
<p>If you do not have access to a cluster computer, AWS is a great option for running QIIME 2. With AWS, you rent computer resources from Amazon at an hourly rate. This is very similar to running on a cluster computer, except that Amazon owns and maintains the hardware (rather than you or your institution owning and maintaining it). If you expect to run QIIME 2 analyses fairly infrequently (e.g., monthly or less) this can be a very cost effective option. You may also qualify for <a class="reference external" href="https://aws.amazon.com/grants/">grants of cloud resource time from Amazon</a>. I received a couple of generous grants from AWS when I first started exploring it for running QIIME 2.</p>
<hr class="docutils" />
<div class="tip admonition">
<p class="admonition-title">How I run QIIME 2</p>
<p>I personally find it convenient to use both a cluster computer and my personal computer for running QIIME 2. My university has a cluster computer that all researchers at the university have access to, and our high performance computing team keeps an up-to-date version of QIIME 2 installed on that computer. I typically run the long-running steps of my workflows on the cluster and download the results to my personal computer. Then, when I’m at a more iterative stage of my analysis (for example, when generating visualizations and running statistical tests), I’ll run those commands locally so I can easily view the results. Most personal computers are powerful enough for these steps.</p>
</div>
</div>
<div class="section" id="installing-qiime-2">
<h3><span class="section-number">1.1.4. </span>Installing QIIME 2<a class="headerlink" href="#installing-qiime-2" title="Permalink to this headline">¶</a></h3>
<p>Because installation instructions for QIIME 2 will change slightly between versions, these instructions are better suited for the QIIME 2 website than for a book. To get the latest installation instructions, see the <a class="reference external" href="https://docs.qiime2.org/2020.8/install/native/">QIIME 2 installation instructions web page</a>. If you have trouble installing QIIME 2, or have questions related to where or how to deploy QIIME 2, post to the <a class="reference external" href="https://forum.qiime2.org">QIIME 2 Forum</a>.</p>
<p>I recommend taking some time now to get QIIME 2 installed on your personal computer, or finding another way to run it so you can follow along with the examples in this book. The data sets that we’ll use as we work through the examples in this book are designed to run quickly on a personal computer, so you won’t need cluster or cloud access to learn from this book. If you’re preparing to run a bigger analysis soon, this is probably a good time to start figuring out what computer you’ll use for that work as it can take some time to work with your high performance computing office to get QIIME 2 installed on their cluster, or to get approval from your institution to bill AWS expenses to them.</p>
</div>
</div>
<div class="section" id="qiime-2-archives">
<h2><span class="section-number">1.2. </span>QIIME 2 archives<a class="headerlink" href="#qiime-2-archives" title="Permalink to this headline">¶</a></h2>
<p>One of the first things that new QIIME 2 users often notice is the <code class="docutils literal notranslate"><span class="pre">.qza</span></code> and <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> files that QIIME 2 uses. All files generated by QIIME 2 are either <code class="docutils literal notranslate"><span class="pre">.qza</span></code> or <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> files, and these are simply zip files that store your data alongside some QIIME 2-specific metadata. You can unzip these files with any typical unzip utility, such as WinZip, 7Zip, or unzip, and you don’t need to have QIIME 2 installed to do that.</p>
<p>For example, let’s download a <code class="docutils literal notranslate"><span class="pre">.qza</span></code> file and take a quick look.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ wget https://some/url/to/sequences.qza
$ unzip sequences.qza
...
</pre></div>
</div>
<p>There are a few things I want you to notice from this example. First, we didn’t use any QIIME 2 commands. We just unzipped the file as we would unzip any zip file. Next, you’ll see there are two top-level directories in the output: <code class="docutils literal notranslate"><span class="pre">data</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata</span></code>. The <code class="docutils literal notranslate"><span class="pre">data</span></code> directory contains a single file, <code class="docutils literal notranslate"><span class="pre">sequences.fasta</span></code>, which contains (you guessed it!) sequence data in fasta format.  If, for example, you’re interested in getting some sequence data out of QIIME 2 to analyze it with another program, you can unzip your <code class="docutils literal notranslate"><span class="pre">.qza</span></code> file, and use the <code class="docutils literal notranslate"><span class="pre">sequences.fasta</span></code> file for what ever you need to do with it. Allowing users to access their data <em>without</em> QIIME 2 was one of the earliest design goals of the system. This ensures that if QIIME 2 isn’t available to you for some reason, you can still access any data that you generated with QIIME 2. You might wonder why we bothered with having QIIME 2 create these zip files in the first place, rather than just have it use the typical file formats like fasta, newick, biom, and so on. That has to do with the other information stored in the zip file - specifically the many files contained in the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> directory. The files in the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> directory are not intended to be viewed by a human, and you don’t need any of them to work with the file (or files) in the <code class="docutils literal notranslate"><span class="pre">data</span></code> directory. QIIME 2 uses the information in the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> directory to record data provenance, helping you to ensure that your bioinformatics work will be reproducible, to store a unique identifier for the data which faciliates data management, and to record the semantic type of the data (we’ll come back to that shortly). This metadata empowers your bioinformatics work in ways that will be more clear as you get further along in your learning. We’ll revisit this topic throughout the book.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">.qza</span></code> file extension is an abbreviation for QIIME Zipped Artifact, and the <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> file extension is an abbreviation for QIIME Zipped Visualization. <code class="docutils literal notranslate"><span class="pre">.qza</span></code> files (which are often simply referred to as <em>artifacts</em>) are intermediary files in a QIIME 2 analysis, usually containing raw data of some sort. These files are generated by QIIME 2 and are intended to be consummed by QIIME 2. `.qzv`` files (which are often simply referred to as <em>visualizations</em>) are terminal results in a QIIME 2 analysis, such as an interactive figure or the results of a statistical test. These files are generated by QIIME 2 and are intended to be consummed by a human.</p>
<p>QIIME 2 also provides some of its own utilities for getting data out of <code class="docutils literal notranslate"><span class="pre">.qza</span></code> and <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> files. If you’re working with the QIIME 2 command line interface (which we’ll use a lot in this book), the most relevant command is <code class="docutils literal notranslate"><span class="pre">qiime</span> <span class="pre">tools</span> <span class="pre">export</span></code>. If you were to run this on the <code class="docutils literal notranslate"><span class="pre">.qza</span></code> file we downloaded above, you’d see the following:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ qiime tools <span class="nb">export</span> --input-path sequences.qza --output-path exported-sequences/
</pre></div>
</div>
<p>This command will unzip the archive, and take all of the files from the <code class="docutils literal notranslate"><span class="pre">data</span></code> directory and place them in <code class="docutils literal notranslate"><span class="pre">exported-sequences</span></code>. Thus if you do have QIIME 2 installed, you can get your data out of a QIIME 2 artifact without all of the QIIME 2-specific metadata using this command.</p>
<div class="note admonition">
<p class="admonition-title">Confused by the term “artifact”?</p>
<p>It has been brought to our attention that the term <em>artifact</em> can be confusing, since it is often used in science to indicate a feature that is not present naturally in a system but rather observed as a result of some technical aspect of studying that system. For example, homopolyer runs such as the <code class="docutils literal notranslate"><span class="pre">A</span></code>s in <code class="docutils literal notranslate"><span class="pre">ACTGTACTAAAAAAAAAAATGCACGTGAC</span></code> were commonly reported by some early sequencing instruments to be longer then they were in nature due to the way the sequencing reaction worked. In QIIME 2 we use the defintion of an artifact as an object that was created by some process, like an archaeological artifact. This is common in data science, and we didn’t realize the potential for confusion until we were a little too far along to easily change the name.</p>
</div>
</div>
<div class="section" id="plugins-and-actions">
<h2><span class="section-number">1.3. </span>Plugins and actions<a class="headerlink" href="#plugins-and-actions" title="Permalink to this headline">¶</a></h2>
<p>People generally think of QIIME 2 as a microbiome bioinformatics system, but the truth is that it’s a lot more general purpose than that. QIIME 2 is built using a plugin architecture. There is a core system, which we call the <em>QIIME 2 Framework</em> or just <em>The Framework</em>. This handles a lot of the behind-the-scenes work, like tracking data provenance and building <code class="docutils literal notranslate"><span class="pre">.qza</span></code> and <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> files. There is no microbiome-specific functionality (or even bioinformatics-specific functionality) in the QIIME 2 Framework. All of the analysis functionality comes in the form of plugins to the framework. There’s only a few things that you need to know about this right now. First, your deployment of QIIME 2 will have some collection of plugins installed. Plugins define actions, which are steps in an analysis workflow. For example, the <code class="docutils literal notranslate"><span class="pre">q2-diversity</span></code> QIIME 2 plugin defines actions including <code class="docutils literal notranslate"><span class="pre">alpha-phylogenetic</span></code> and <code class="docutils literal notranslate"><span class="pre">beta-phylogenetic</span></code> which can apply phylogenetic alpha and beta diversity metrics, respectively, to your data. If you don’t have the <code class="docutils literal notranslate"><span class="pre">q2-diversity</span></code> plugin installed, you won’t have access to those actions. To find out what QIIME 2 plugins you currently have installed, you can run the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ qiime --help
</pre></div>
</div>
<p>If you want to see what actions are defined by a plugin, you can call <code class="docutils literal notranslate"><span class="pre">--help</span></code> on that plugin. For example, to see what actions are available from the <code class="docutils literal notranslate"><span class="pre">q2-diversity</span></code> plugin, you can run the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ qiime diversity --help
</pre></div>
</div>
<p>You should see <code class="docutils literal notranslate"><span class="pre">alpha-phylogenetic</span></code> and <code class="docutils literal notranslate"><span class="pre">beta-phylogenetic</span></code> in that list, among other actions. You could go one step further if you’d like to learn about how to use those actions by calling help on the action. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ qiime diversity alpha-phylogenetic --help
</pre></div>
</div>
<p>The output that you get from that might look a little mysterious right now. When you finish Part 1 of this book, you’ll understand how to read that help text and use it learn how to use QIIME 2 actions you’ve never used before.</p>
<p>Another thing to know about plugins is that anyone can create and distribute them. For example, if a graduate student develops some new analysis functionality that they want to use with QIIME 2, that can create their own QIIME 2 plugin. If they want others to be able to use it, they can distribute that plugin. The <a class="reference external" href="https://library.qiime2.org">QIIME 2 Library</a> is a website developed by the QIIME 2 team to help with dissemination of plugins. It’s a great site to visit if you want to discover new analysis functionality. Developing and disseminating plugins is covered in Part 3 of this book.</p>
<p>QIIME 2 actions come in three varieties, as of this writing. Methods are a type of QIIME 2 action that generate one or more <code class="docutils literal notranslate"><span class="pre">.qza</span></code> files as output. Since <code class="docutils literal notranslate"><span class="pre">.qza</span></code> files are intermediary results in QIIME 2 (as discussed in the previous section), Methods typically represent some sort of processing step in your analysis, such as taxonomic annotation of sequences. The <code class="docutils literal notranslate"><span class="pre">alpha-phylogenetic</span></code> and <code class="docutils literal notranslate"><span class="pre">beta-phylogenetic</span></code> actions described above are Methods. Visualizers are a type of QIIME 2 action that generate one or more <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> files as output. You’ll remember that <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> files are terminal results in QIIME 2, so these are steps that terminal in a workflow. An example of a QIIME 2 visualizer is the <code class="docutils literal notranslate"><span class="pre">beta-group-significance</span></code> action in the <code class="docutils literal notranslate"><span class="pre">q2-diversity</span></code> plugin, which runs PERMANOVA, ANOSIM, or PERMDISP on your data, and reports the result of the statistical test. The third type of action in QIIME 2 is a Pipeline, which can generate one or more <code class="docutils literal notranslate"><span class="pre">.qza</span></code> and/or <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> as output. <code class="docutils literal notranslate"><span class="pre">Pipelines</span></code> are special in that they’re a type of action that can call other actions. They are often used by developers to define simplify common workflows so they can be run by users in a single step. For example, the <code class="docutils literal notranslate"><span class="pre">core-metrics-phylogenetic</span></code> action in the <code class="docutils literal notranslate"><span class="pre">q2-diversity</span></code> plugin is a Pipeline that runs both <code class="docutils literal notranslate"><span class="pre">alpha-phylogenetic</span></code> and <code class="docutils literal notranslate"><span class="pre">beta-phylogenetic</span></code>, as well as several other actions, in a single command. In total it runs about 20 different actions, so it saves a lot of typing to be able to do achieve that with a single command.</p>
<p>There’s more to know about plugins and actions in QIIME 2, but this will get you started.</p>
</div>
<div class="section" id="semantic-types">
<h2><span class="section-number">1.4. </span>Semantic types<a class="headerlink" href="#semantic-types" title="Permalink to this headline">¶</a></h2>
<p>The next topic that should be briefly covered before you start using QIIME 2 is the notion of types in QIIME 2. The term <em>type</em> is
overloaded with a few different concepts, so I’ll start by talking about two ways that it’s commonly used, and then introduce a third way that it’s used less frequently but which is important to QIIME 2 (and which could help other systems, in my opinion). By disambiguating this concept now I think we’ll avoid confusion later, and you’ll be in a better place to understand QIIME 2 help text and other documentation.</p>
<div class="margin sidebar">
<p class="sidebar-title"></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference external" href="https://www.youtube.com/watch?v=PUsvtJgpNtE">This video</a> on the QIIME 2 YouTube channel discusses semantic types.</p>
</div>
</div>
<p>The three kinds of types that are used in QIIME 2 are file types, data types, and semantic types. File types (or formats) refer to what you probably think of when you hear that phrase: the format of a file used to store some data. For example, newick is a file type that is used for storing phylogenetic trees. Files are used most commonly for archiving data when it’s not actively in use. Data types refer to how data is represented in a computer’s memory (i.e., RAM) while it’s actively in use. For example, if you are adding a root to an unrooted phylogenetic tree (a concept discussed in Part 2 of this book), you may use a tool like IQTree2. You would provide a path to the file containing the unrooted phylogenetic tree to IQTree2, and IQTree2 would load that tree into some <em>data structure</em> in the computer’s memory to work on it. The data structure or type, that IQTree2 uses internally to represent the phylogenetic tree will be a decision made by the developers of IQTree2. If it successfully completes the requested rooting operation, IQTree2 would write the new tree from an internal data type into a new newick-formatted file on the hard disk, and exit. As a software user, you shouldn’t need to know or care about what data types are used internally by a program - you just care about what file types are used as input and output. Computer programmers care a lot about internal data types: choosing an appropriate one has huge impacts on the software.</p>
<p>The third <em>type</em> that is important in QIIME 2 is the semantic type of data. This is a representation of the <em>meaning</em> of the data, which is not necessarily represented by either a file type or a data type. For example, two semantic types used in QIIME 2 are <code class="docutils literal notranslate"><span class="pre">Phylogeny[Rooted]</span></code> and <code class="docutils literal notranslate"><span class="pre">Phylogeny[Unrooted]</span></code>, which are used to represent rooted and unrooted trees, respectively. Both rooted and unrooted trees are commonly stored in newick files, and a computer program needs to parse (i.e., load data from a file into a in-memory data structure) to know if a tree is rooted or unrooted. For large trees, this can be a slow operation. There are some operations, such as rooting a tree, that only make sense to perform on unrooted trees. So, if you have a very large tree that you want to root, you may provide a newick file to a program that will perform that rooting. If you accidentally provide a rooted tree (say because you have tried rooting it with a few different approaches that you want to evaluate), it may take the program some time to parse the file (say 20 minutes) after which it may fail if it discovers that the tree is already rooted. That sort of delayed notification can be very frustrating as a user, since it’s easily missed until a lot of time has passed. I often will start a long-running command on my university cluster computer just before the weekend. I’ll typically check on the job for a few minutes, to make sure that it seems to be starting ok. I may then leave, in the hope that the job completes over the weekend and I’ll have data to work with on Monday morning. It’s very frustrating to come in Monday morning and find out that my job failed just a few minutes after I left on Friday for a reason that I could have quickly addressed had I known in time.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>There’s actually a worse outcome than a delayed error from a computer program when inappropriate input is provided. When a program fails and provides an error message to the user, whether or not that error message helps the user solve the problem, the program has failed loudly. Something went wrong, and it told the user about it. The program could instead fail quietly. This might happen if the program doesn’t realize the input the user provided is in appropriate (e.g., an already rooted tree is provided to a program that roots an unrooted phylogenetic tree), and it runs the rooted tree through its algorithm, misinterprets something because it was provided with the wrong input, and generates an incorrect rooted tree as a result. Quiet failures can be very difficult or impossible for a user to detect, because it looks like everything has worked as expected. Failing quietly is thus <em>much</em> worse than failing loudly - it could waste many hours of your time, and could even lead to you publishing invalid findings.</p>
</div>
<p>QIIME 2 semantic types help with this, because they provide information on what the data in a QIIME 2 <code class="docutils literal notranslate"><span class="pre">.qza</span></code> file means without having to parse anything in the <code class="docutils literal notranslate"><span class="pre">data</span></code> directory. All QIIME 2 artifacts have a semantic type associated with them (it’s one of the pieces of information stored in the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> directory), and QIIME 2 methods will describe what semantic types they take as input(s), and what semantic types they generate as output(s). For example, the <code class="docutils literal notranslate"><span class="pre">q2-phylogeny</span></code> plugin defines a method called <code class="docutils literal notranslate"><span class="pre">midpoint_root</span></code>. Call help on this method using the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ qiime phylogeny midpoint-root --help
</pre></div>
</div>
<p>You can see from the resulting help text that this method takes one input, an artifact of semantic type <code class="docutils literal notranslate"><span class="pre">Phylogeny[Unrooted]</span></code>. It also generates one output, an artifact of semantic type <code class="docutils literal notranslate"><span class="pre">Phylogeny[Rooted]</span></code>. This makes intuitive sense: an action that adds a root to a phylogenetic tree (as described in the help text for this method) takes an unrooted tree as input and generates a rooted tree as output.</p>
<p>There is a many-to-many relationship between file types, data types, and semantic types. It’s possible that a given semantic type could be represented on disk by different file types. That’s well exemplified by the many different formats that are used to store demultiplexed sequence and sequence quality data. For example, this may be in one a few variants of the fastq format, or in the fasta/qual format. Additionally, data from multiple samples may be contained in one single file or split into per-sample files. Regardless of which of these file formats the data is stored in, QIIME 2 will assign the same semantic type (in this case, <code class="docutils literal notranslate"><span class="pre">SampleData[SequencesWithQuality]</span></code>. Similarly, the data type used in memory might differ depending on what operations are to be performed on the data, or based on the preference of the programmer. QIIME 2 use the semantic type <code class="docutils literal notranslate"><span class="pre">FeatureTable[Frequency]</span></code> to represent the idea of a feature table that contains counts of features (e.g., bacterial genera) on a per sample basis. Many different actions can be applied to <code class="docutils literal notranslate"><span class="pre">FeatureTable[Frequency]</span></code> artifacts in QIIME 2. When a plugin developer defines a new action that takes a <code class="docutils literal notranslate"><span class="pre">FeatureTable[Frequency]</span></code> as input, they can choose whether to load the table into a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> or <code class="docutils literal notranslate"><span class="pre">biom.Table</span></code> object, which are two different data types.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>That last paragraph was a bit technical. Don’t worry if you got lost in the details - just take away the idea that there is not a one-to-one relationship between file types, data types, and semantic types in QIIME 2. Each kind of type represents different information about the data.</p>
<p>The motivation for creating QIIME 2’s semantic type system was to avoid issues that can arise from providing inappropriate data to actions. The semantic type system also helps users and developers better understand the intent of QIIME 2 actions by assigning meaning to the input and output, and allows for the discovery of new potentially relevant QIIME 2 actions (more on this later). Throughout the next few chapters I’ll point out semantic types of some inputs and outputs as we come across them. Again, there’s more to know on this topic, but that learning can be deferred until its needed.</p>
</div>
</div>
<div class="section" id="qiime-2-view">
<h2><span class="section-number">1.5. </span>QIIME 2 View<a class="headerlink" href="#qiime-2-view" title="Permalink to this headline">¶</a></h2>
<p>QIIME 2 View is a web-based viewer for <code class="docutils literal notranslate"><span class="pre">.qza</span></code> and <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> files. If you’ve never visited QIIME 2 View, take a minute to <a class="reference external" href="https://view.qiime2.org">go to the website</a> now. This site allows for you to view QIIME 2 results on computers that don’t have QIIME 2 installed on them, and there are a few examples that you can look at in the gallery on that page. I use QIIME 2 View daily, for a few different situations. First, if I’m running analyses on a cluster computer than doesn’t provide a graphical interface, it’s a convenient way to view those results without having to load QIIME 2 on another computer. If I have a copy of a <code class="docutils literal notranslate"><span class="pre">.qza</span></code> or <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> file on my local computer (e.g., if I copied it over from the cluster) I can navigate to QIIME 2 View, and drag-and-drop the file on the QIIME 2 View page to look at it. Another scenario where this is helpful is if when I’m sharing interactive QIIME 2 results with someone who doesn’t have QIIME 2 installed, for example a collaborator. I can send them <code class="docutils literal notranslate"><span class="pre">.qzv</span></code> files by email (or use the Dropbox sharing option on QIIME 2 View), and they can load and interact with the results on their computer.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>When you’re using QIIME 2 View, your data isn’t uploaded to a server. Rather the website acts as an application launcher that allows you to view local files. This means that you don’t need to be concerned about data privacy issues with QIIME 2 View - the data never leaves your computer.</p>
</div>
</div>
<div class="section" id="getting-started">
<h2><span class="section-number">1.6. </span>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h2>
<p>Ok, that’s enough discussion about QIIME 2 for now. It’s time to start using it. Don’t worry if you feel like you don’t fully understand some of the technical details I covered in this chapter right now. My goal was to introduce these ideas to you here, and we’ll revisit them in the remaining chapters in the book. As we contextualize these ideas by using QIIME 2, they’ll become clearer.</p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./using"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="../front-matter/preface.html" title="previous page">Preface</a>
    <a class='right-next' id="next-link" href="overview-tutorial.html" title="next page"><span class="section-number">2. </span>An interactive overview of QIIME 2</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By J Gregory Caporaso<br/>
        
            &copy; Copyright 2020-2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../_static/js/index.3da636dd464baa7582d2.js"></script>


    
  </body>
</html>